### PR TITLE
✨ Generate event specific reports

### PR DESCRIPTION
**Pull request by ubmarco**
This PR makes it possible to aggregate the runtime of event listeners for all fired Sphinx events.
Tricky part in pyinstrument is that all event listeners appear under `EventManager.emit`, so the event name is lost.
The PR monkeypatches Sphinx to use a custom Event listener which injects an event specific `emit` function, so it can be tracked in pyinstrument.

The PR can be tested with `sphinx-analysis --project needs --pyinstrument --tree`.
A new JSON report of name `pyinstrument_sphinx_events.json` is created.
It contains all unique combinations of file path, class and function of all event listeners, so it will pick up all extensions that have listeners registered.
The dictionary `sphinx_performance/sphinx_events.py:CUSTOM_FRAMES_BY_EVENT` can be modified to add more frames of interest (unrelated to events). It is currently used to track the runtime of `sphinxcontrib-plantuml`.

The generated JSON file can be used for 2 purposes:
- As a support to the pyinstrument HTML report to get a per-event aggregation
- To run in a CI and check how a PR to Sphinx or an extension affects performance

The HTML report now contains event specific function names, so event names can be searched.

### CODE REVIEW

**Comment by danwos** [comment]
I get an error during report generation when I run `sphinx-analysis --project needs --pyinstrument --tree`:

```
  File "/home/daniel/workspace/sphinx/sphinx-performance/sphinx_performance/analysis.py", line 268, in cli_analysis
    with Path.open("pyinstrument_profile.html", "w") as events_json_file:
  File "/usr/lib/python3.10/pathlib.py", line 1119, in open
    return self._accessor.open(self, mode, buffering, encoding, errors,
AttributeError: 'str' object has no attribute '_accessor'
```

**Comment by ubmarco** [comment]
I can confirm this bug for Python 3.10. For Python 3.11 it works. Need to check.

File: sphinx_performance/analysis.py [hunk 1]
```diff
@@ -251,16 +253,29 @@ def cli_analysis(
...
         )
-        with Path.open("pyinstrument_profile.html", "w") as profile_html:
-            profile_html.write(html_data)
+        html_data = HTMLRendererFromJson(
+            show_all=show_all,
+            processor_options=processor_options,
+        ).render(
+            json_str,
+        )
+        with Path.open(Path("pyinstrument_profile.html"), "w") as events_json_file:
```
Comment (chrisjsewell) [inline sphinx_performance/analysis.py:268]:
```suggestion
        with Path("pyinstrument_profile.html").open("w") as events_json_file:
```

Comment (chrisjsewell) [reply sphinx_performance/analysis.py:268]: Also this path should probably be absolute, otherwise you are relying on the current context?

**Review by chrisjsewell — CHANGES_REQUESTED** [review]

File: sphinx_performance/analysis.py (hunk 1 above)
Comment (ubmarco) [reply sphinx_performance/analysis.py]: we have 2 options
- integration to /tmp dir where the build happens
- store the CWD
Of those 2 I prefer CWD as /tmp has to be located first and it changes all the tiem.

**Review by chrisjsewell — COMMENTED** [review]
Cheers, yeh I guess ideally it would be nice to have some tests 😅
But as there were already no tests for this package, I won't block on it

**Review by chrisjsewell — APPROVED** [review]

**Review by danwos — APPROVED** [review]

**Pull request merged**